pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	registry := reference.Domain(named)
+	if strings.ContainsAny(registry, ".:") || registry == "localhost" {
+		return false, ref, nil
+	}
+
+	return true, ref, nil
+
+}
+
+// Add records the specified name-value pair as a new short-name alias to the
+// user-specific aliases.conf.  It may override an existing alias for `name`.
+//
+// Note that tags are stripped from the specified name.
+func Add(ctx *types.SystemContext, name string, value string) error {
+	isShort, shortRef, err := isShortName(name)
+	if err != nil {
+		return err
+	}
+	if !isShort {
+		return errors.Errorf("%q is not a short name", name)
+	}
+	normalizedShortName := name
+	if tagged, isTagged := shortRef.(reference.Tagged); isTagged {
+		normalizedShortName = strings.TrimSuffix(name, ":"+tagged.Tag())
+	}
+	return sysregistriesv2.AddShortNameAlias(ctx, normalizedShortName, value)
+}
+
+// Remove clears the short-name alias for the specified name.  It throws an
+// error in case name does not exist in the machine-generated
+// short-name-alias.conf.  In such case, the alias must be specified in one of
+// the registries.conf files, which is the users' responsibility.
+//
+// Note that tags are stripped from the specified name.
+func Remove(ctx *types.SystemContext, name string) error {
+	isShort, shortRef, err := isShortName(name)
+	if err != nil {
+		return err
+	}
+	if !isShort {
+		return errors.Errorf("%q is not a short name", name)
+	}
+	normalizedShortName := name
+	if tagged, isTagged := shortRef.(reference.Tagged); isTagged {
+		normalizedShortName = strings.TrimSuffix(name, ":"+tagged.Tag())
+	}
+	return sysregistriesv2.RemoveShortNameAlias(ctx, normalizedShortName)
+}
+
+// Resolve resolves the specified name to either one or more fully-qualified
+// image references.  If the specified name is already a fully-qualified
+// reference (i.e., not a short name), it is returned as is.  In case, it's a
+// short name, it's resolved according to the ShortNameMode in the
+// SystemContext (is specified) or in the registries.conf.
+//
+// Note that returned slice may be empty if no matching alias and no
+// unqualified-search registries could be found.  Names including a digest are
+// returned as is.  Tags are stripped from the specified name.
+func Resolve(ctx *types.SystemContext, name string) ([]reference.Named, error) {
+	isShort, shortRef, err := isShortName(name)
+	if err != nil {
+		return nil, err
+	}
+	if !isShort { // no short name
+		named, isNamed := shortRef.(reference.Named)
+		if !isNamed {
+			return nil, errors.Wrapf(err, "name %q is neither a short name nor a valid named reference", name)
+		}
+		return []reference.Named{named}, nil
+	}
+
+	// Query the registry for unqualified-search registries.
+	unqualifiedSearchRegistries, err := sysregistriesv2.UnqualifiedSearchRegistries(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Detect which mode we're running in.
+	var mode types.ShortNameMode
+	if ctx != nil && ctx.ShortNameMode != nil {
```

Comment:
I’d expect `GetShortNameMode` to handle this, so that it is the same in all callers of that function.